File: ConvertANumberToHexadecimal/convert_a_number_to_hexadecimal.dart
/*

-* 405. Convert a Number to Hexadecimal *-

Given an integer num, return a string representing its hexadecimal representation. For negative integers, two’s complement method is used.

All the letters in the answer string should be lowercase characters, and there should not be any leading zeros in the answer except for the zero itself.

Note: You are not allowed to use any built-in library method to directly solve this problem.

 

Example 1:

Input: num = 26
Output: "1a"
Example 2:

Input: num = -1
Output: "ffffffff"
 

Constraints:

-231 <= num <= 231 - 1


*/
String toHexString(int num) {
  if (num == 0) {
    return "0";
  }

  int n = num.toUnsigned(32);
  final String ref = "0123456789abcdef";

  String result = "";

  while (n != 0) {
    result = ref[n & 0xF] + result;
    n >>= 4;
  }

  return result;
}

String toHex(int num) {
  return (num < 0)
      ? (num.toUnsigned(32)).toRadixString(16)
      : num.toRadixString(16);
}

extension ToHex on int {
  String toHexString(int num) {
    if (num == 0) {
      return "0";
    }

    int n = num.toUnsigned(32);
    final String ref = "0123456789abcdef";

    String result = "";

    while (n != 0) {
      result = ref[n & 0xF] + result;
      n >>= 4;
    }

    return result;
  }
}

File: ConvertANumberToHexadecimal/convert_a_number_to_hexadecimal.go
package main

func toHex(num int) string {
	if num == 0 {
		return "0"
	}

	const bitsPerDigit = 4
	const mask = (1 << bitsPerDigit) - 1

	hex := make([]byte, 8)
	i := 7

	// Convert negative numbers to their two's complement representation
	if num < 0 {
		num = (1 << 32) + num
	}

	for num != 0 {
		nib := num & mask
		hex[i] = getHexDigit(nib)
		num >>= bitsPerDigit
		i--
	}

	return string(hex[i+1:])
}

func getHexDigit(n int) byte {
	if n < 10 {
		return byte('0' + n)
	}
	return byte('a' + n - 10)
}

//--------------------------

